// @jupyter-server/resource-usage/lib/indicator.js
import React, { useState } from 'react';
import { Sparklines, SparklinesLine, SparklinesSpots } from 'react-sparklines';
/**
 * A indicator fill component.
 */
const IndicatorFiller = ({ percentage, color, }) => {
    return (React.createElement("div", { className: "jp-IndicatorFiller", style: {
            width: `${percentage * 100}%`,
            background: `${color}`,
        } }));
};
/**
 * An indicator bar component.
 */
const IndicatorBar = ({ values, percentage, baseColor, }) => {
    const [isSparklines, setIsSparklines] = useState(false);
    const toggleSparklines = () => {
        setIsSparklines(!isSparklines);
    };
    const color = percentage > 0.5 ? (percentage > 0.8 ? 'red' : 'orange') : baseColor;
    return (React.createElement("div", { className: "jp-IndicatorBar", onClick: () => toggleSparklines() },
        isSparklines && (React.createElement(Sparklines, { data: values, min: 0.0, max: 1.0, limit: values.length, margin: 0 },
            React.createElement(SparklinesLine, { style: {
                    stroke: color,
                    strokeWidth: 4,
                    fill: color,
                    fillOpacity: 1,
                } }),
            React.createElement(SparklinesSpots, null))),
        !isSparklines && (React.createElement(IndicatorFiller, { percentage: percentage, color: color }))));
};
/**
 * An incicator component for displaying resource usage.
 *
 */
export const IndicatorComponent = ({ enabled, values, label, color, text, }) => {
    const percentage = values[values.length - 1];
    return (React.createElement(React.Fragment, null, enabled && (React.createElement("div", { className: "jp-IndicatorContainer" },
        React.createElement("div", { className: "jp-IndicatorText" }, label),
        percentage !== null && (React.createElement("div", { className: "jp-IndicatorWrapper" },
            React.createElement(IndicatorBar, { values: values, percentage: percentage, baseColor: color }))),
        React.createElement("div", { className: "jp-IndicatorText" }, text)))));
};
